### PR TITLE
Separate rights statement portions into different objects

### PR DESCRIPTION
**Pull request by p-galligan**
Separates objects out from `base.json`. Rewrites the objects to clarify the property names and sections.

I have not yet removed any data from the `base.json` file. The rights statement and basis-specific sections are much closer to PREMIS now. 

You'll also see that there's some conditional logic happening at the end of the rights statement and rights granted objects that gives us more control over what we're allowing to show up when.

A couple of questions:

1. I have not replicated the note object for the notes in these objects because they are generally much simpler than ArchivesSpace notes. Anyone see any issues with this?
2. What exactly is the source_rights_statement doing here? I haven't made any changes to it, but it looks a lot like a duplication of the rights statement and granted objects.

### CODE REVIEW

**Comment by bonniegee** [comment]
Thanks @p-galligan! Yes, the source_rights_statements (from looking at the blame) was part of adding the rights schema; the associated PR is #9. And yes, the `note` here is sufficient.

I see that dates are included for each basis type, with a basis specific label. We're not currently doing that with applications that produce these rights statements, and are just using generic start and end dates. This is not what PREMIS calls for, but that specific labeling happens further down the line (like in Archivematica). I'm thinking we can stick with that approach; what are folks' thoughts?

**Comment by p-galligan** [comment]
> This is not what PREMIS calls for, but that specific labeling happens further down the line (like in Archivematica). I'm thinking we can stick with that approach; what are folks' thoughts?

I'm happy to pull the start and end dates out of the individual sections and leave it in the base `rights_statement.json` file, especially since there's basically no change to these properties across the different objects.

**Comment by p-galligan** [comment]
I've also added properties for identifiers, and we don't use those, but it looked like we might want to in the future in the RAC documentation.

File: rac_schemas/schemas/rights_copyright.json [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      "end_date": {
+        "$id": "#/rights_copyright/properties/end_date",
+        "type": ["string", "null"],
+        "title": "Copyright End Date",
+        "examples": [
+          "2019-08-23",
+          "open"
+        ]
+      },
+      "copyright_note": {
```
Comment (helrond) [inline rac_schemas/schemas/rights_copyright.json]: Arguably if we're going to pull dates out of these bases and into the more generic rights_basis.json file, the same could be done for notes.

File: rac_schemas/schemas/rights_copyright.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        "$id": "#/rights_copyright/properties/copyright_note",
+        "type": "array",
+        "title": "Copyright Note",
+        "items": {
+          "type": "string",
+          "maxLength": 2500
```
Comment (helrond) [inline rac_schemas/schemas/rights_copyright.json]: Where is this number coming from?

File: rac_schemas/schemas/rights_granted.json [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        "title": "Rights Granted Restriction Statement",
+        "enum": [
+          "allow",
+					"conditional",
+          "disallow"
+        ],
+        "examples": [
+          "allow"
+        ],
+        "pattern": "^(.*)$"
```
Comment (helrond) [inline rac_schemas/schemas/rights_granted.json]: I believe this is redundant on a field which has an `enum`?

File: rac_schemas/schemas/rights_granted.json [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+				"properties": { "grant_restriction": { "const": "conditional" } }
+			},
+			"then": {
+				"required": [ "act", "grant_restriction", "granted_note" ]
+			},
+			"else": {
+				"required": [ "act", "grant_restriction" ]
+			}
+		}
+  }]
```
Comment (helrond) [inline rac_schemas/schemas/rights_granted.json]: When we write tests, let's make sure we're testing this logic.

File: rac_schemas/schemas/rights_license.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    "$id": "#/rights_license",
+    "type": "object",
+		"title": "PREMIS License Information",
+    "properties": {
+      "license_identifier": {
+        "$id": "#/rights_license/properties/license_identifier",
+        "type": "string",
+        "title": "License Documentation Identifier",
+				"minLength": 1,
+        "maxLength": 250
```
Comment (helrond) [inline rac_schemas/schemas/rights_license.json]: Where is this length coming from?

File: rac_schemas/schemas/rights_license.json [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      },
+      "license_terms": {
+        "$id": "#/rights_license/properties/license_terms",
+        "type": "string",
+        "title": "License Terms",
+        "maxLength": 2500,
```
Comment (helrond) [inline rac_schemas/schemas/rights_license.json]: Where is this number coming from?

File: rac_schemas/schemas/rights_statement.json [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      },
+      "rights_granted": {
+        "type": "array",
+        "items": {
+          "allOf": [{
+            "$ref": "rights_granted.json#/rights_granted"
+          }]
+        }
+      },
+      "if": {
```
Comment (helrond) [inline rac_schemas/schemas/rights_statement.json]: OK, a couple of things about this:
- Can't we just use `oneOf` here rather than all this conditional logic? https://json-schema.org/understanding-json-schema/reference/combining.html#oneof
- It looks to me like the object this is expecting is something like:
```
{
  "rights_basis": "copyright",
  "rights_granted": "allow",
  "basis_information: {
    ...  //basis information here
  }
}
```

Which...I don't think is what we want?

**Review by helrond — CHANGES_REQUESTED** [review]
Some specific questions below, but a couple of things across the board which need to be addressed:
- The spacing in these files is inconsistent, and makes them hard to read. Can we format them all so they have matching indents, etc.?
- Schema titles need to be updated for all files.
- the top-level `$id` field is the same for all of the rights bases (`rights_statement.json`) and needs to be updated.
- I'm not sure why we're using `allOf` in the rights basis and rights granted definition files?
- There's an invalid JSON file somewhere in here. I think it might be `rights_granted.json`

**Comment by helrond** [comment]
One more comment here about the role of `rights_statement.json` versus `source_rights_statement.json` - the original idea was that there might be some distinction between rights statements pre-Pisces and post-Pisces. My sense is that is no longer true. For one thing, we're not transforming rights statements in Pisces. But even if we were doing that, it seems to make more sense to just structure rights statements the way we want to in Aurora and Aquila, so that a "transformation" would just be a copy. 

So, bottom line, I'd suggest getting rid of the `source_rights_statement` definition, and pointing to the `rights_statement` definition in the `digitization_bag` and `legacy_digital_bag` definitions.

File: rac_schemas/schemas/rights_license.json (hunk 5 above)
Comment (p-galligan) [reply rac_schemas/schemas/rights_license.json]: Arbitrarily chosen, but I can take it out. Just figured we'd want some upper limit on an identifier length.

File: rac_schemas/schemas/rights_copyright.json (hunk 2 above)
Comment (p-galligan) [reply rac_schemas/schemas/rights_copyright.json]: Generally arbitrarily chosen, though roughly half the size of a notes length in AS. However, I can remove the limit.

File: rac_schemas/schemas/rights_license.json (hunk 6 above)
Comment (p-galligan) [reply rac_schemas/schemas/rights_license.json]: Again, arbitrarily chosen high number. happy to remove an limits.

File: rac_schemas/schemas/rights_copyright.json [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+		},
+		"copyright_jurisdiction": {
+			"$id": "#/properties/copyright_jurisdiction",
+			"type": "string",
+			"title": "Copyright Jurisdiction",
+			"minLength": 1,
+			"examples": [
+				"FR"
+			],
+			"pattern": "^[A-Z]{2}$"
```
Comment (bonniegee) [inline rac_schemas/schemas/rights_copyright.json]: I'm pretty sure all our systems use lower case for country codes--does that sound right @helrond ?

**Comment by bonniegee** [comment]
> I've also added properties for identifiers, and we don't use those, but it looked like we might want to in the future in the RAC documentation.

@p-galligan by "identifiers" are you referring to documentation identifiers?

**Comment by p-galligan** [comment]
> > I've also added properties for identifiers, and we don't use those, but it looked like we might want to in the future in the RAC documentation.
> 
> @p-galligan by "identifiers" are you referring to documentation identifiers?

Yes, for the different basis files. The rights statement identifier is the basic identifier you might use in the system.

File: rac_schemas/schemas/rights_copyright.json (hunk 8 above)
Comment (helrond) [reply rac_schemas/schemas/rights_copyright.json]: Well, [this](https://docs.rockarch.org/premis-rights-guidelines/guidelines/#controlled-vocabularies-and-standards) says use ISO 3166, which is uppercase...so if we're using lowercase elsewhere we may be doing it wrong?

Also, reminder that this is a validation against the serialized JSON data, as opposed to the values stored in the database.

**Comment by bonniegee** [comment]
@p-galligan Can you say more about what you mean by "basis files" and "the system"?

**Comment by p-galligan** [comment]
Sorry, I wasn't being completely clear here.

The "basis files" are the specific object schemas for different rights basis choices (rights_other, rights_license, rights_copyright, and rights_statute). Those have identifiers for documentation identifiers.

"The system" was referring to the usage of the Premis rightsStatementIdentifer: "The designation used to identify the Rights statement uniquely within a preservation repository system." So referring to any system that might create an identifier that we want to track for rights statements. Might be helpful in the future for an identifier management system eventually.

**Comment by bonniegee** [comment]
Thanks @p-galligan ! That documentation identifier semantic unit is referring to external documentation, not the rights basis themselves. This is not explained well in the local guidelines because we don't use it, so it's pretty confusing. 

One of the pieces of this that's important but not in the local guidelines (because we don't use it) is that the Documentation Identifier semantic units are containers, that contain three semantic units: Documentation Identifier Type, Documentation Identifier Role, and Documentation Identifier Value. This seems to add a fair amount of complexity for something our systems currently aren't creating, so I'm wondering if we should leave it out for now?

File: rac_schemas/schemas/rights_copyright.json (hunk 8 above)
Comment (bonniegee) [reply rac_schemas/schemas/rights_copyright.json]: Okay--that piece of the local guidelines was definitely modeled off of the Archivematica documentation, which calls for that standard but uses lowercase: https://www.archivematica.org/en/docs/archivematica-1.12/user-manual/metadata/premis/#copyright-jurisdiction 🤷‍♀️

It looks like the existing fixtures we have (like in Ursa Major) are all lowercase, so should we continue with that convention and update the docs?

Comment (helrond) [reply rac_schemas/schemas/rights_copyright.json]: Good times.

OK, it looks like Aurora also produces lowercase as well. So yes, let's update docs to say lowercase and enforce that in this schema.

File: rac_schemas/schemas/rights_statute.json [hunk 9]
```diff
@@ -14,9 +14,9 @@
 			"title": "Statute Jurisdiction",
 			"minLength": 1,
 			"examples": [
-				"EN"
+				"en"
```
Comment (helrond) [inline rac_schemas/schemas/rights_statute.json]: This is a language code, not a country code 😎

Comment (p-galligan) [reply rac_schemas/schemas/rights_statute.json]: Lol, so I want to be clear with @bonniegee : we're using two-letter country codes from ISO 3166, yes? So this would be `us` instead.

**Comment by p-galligan** [comment]
@helrond I'm not sure how to get links to the other JSON schemas to validate correctly in the `rights_statement.json` file. Would you mind taking a look when you get a chance? I think I've probably just structured the refs incorrectly?

**Comment by helrond** [comment]
@p-galligan I'm not 100% sure what's going on here, but I would suggest getting rid of the duplicate definitions in `base.json` as a first step. It's going to be hard to figure things out if those definitions are still there.

**Comment by p-galligan** [comment]
> @p-galligan I'm not 100% sure what's going on here, but I would suggest getting rid of the duplicate definitions in `base.json` as a first step. It's going to be hard to figure things out if those definitions are still there.

Still getting the same error message. However, while updating fixtures, I realized I may have misunderstood something that @bonniegee was asking for earlier? I've abstracted start date, end date, and notes from specific basis files and rights granted, but those still need to be referenced as available for part of the rights_granted section, correct? Otherwise it would just be a single date for each statement, but multiple grant acts with possibly different dates. Correct?

**Comment by helrond** [comment]
> I've abstracted start date, end date, and notes from specific basis files and rights granted, but those still need to be referenced as available for part of the rights_granted section, correct? Otherwise it would just be a single date for each statement, but multiple grant acts with possibly different dates. Correct?

That's right, you'll need to reference those objects from `base.json`.

Resolving the import errors: the way the schemas are currently set up assumes that all common objects will be in `base.json` and will be referenced from there. So, the options are as follows:
- Move `rights_statement` and all associated objects (`rights_granted` and the different bases) into `base.json` and make the current objects just references (see, for example, `group.json`)
- Take a different approach, such as the one advocated in the first answer [here](https://stackoverflow.com/questions/42159346/jsonschema-refresolver-to-resolve-multiple-refs-in-python)

I think I'd lean towards the second option, because I think it will allow us to get rid of the weird structure where we have things in a `base` file and then the more specifically-named JSON files just point to them with references.

**Comment by p-galligan** [comment]
Ok, this is ready for review @bonniegee and @helrond .

There are some properties that have been broken up into some very small objects. And sometimes the different basis objects are only a single property, but that was generally necessary to allow for `oneOf` validation. I've also kept conditional logic to make sure that only the matching basis object is used when a specific rights basis is selected.

File: fixtures/invalid/copyright_basis/no_status.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
+  "copyright_jurisdiction": "US"
```
Comment (helrond) [inline fixtures/invalid/copyright_basis/no_status.json]: lowercase
```suggestion
  "copyright_jurisdiction": "us"
```

File: fixtures/invalid/digitization_bag/4c50bbba-5698-4add-893a-27be921f0345.json [hunk 11]
```diff
@@ -5,43 +5,34 @@
         {
```
Comment (helrond) [inline fixtures/invalid/digitization_bag/4c50bbba-5698-4add-893a-27be921f0345.json]: I would copy and paste a valid rights statement in here, presuming that this is testing something other than whether or not the embedded rights statement is valid. I think we should also rename this fixture so it's more obvious what exactly this is invalidating.

File: fixtures/invalid/digitization_bag/4c50bbba-5698-4add-893a-27be921f0345.json [hunk 12]
```diff
@@ -5,43 +5,34 @@
...
                 }
             ],
             "start_date": "1994-05-14",
-            "status": "copyrighted"
+            "copyright_status": "copyrighted",
+            "copyright_jurisdiction": "us"
         },
         {
             "end_date": "2009-06-22",
             "note": "Records embargoed for 10 years after date of creation.",
```
Comment (helrond) [inline fixtures/invalid/digitization_bag/4c50bbba-5698-4add-893a-27be921f0345.json]: For example, I don't think that `note` is a valid field.

**Review by helrond — CHANGES_REQUESTED** [review]
I've added a few specific comments in here, but a couple of general comments:
- Breaking out `rights_date`, `rights_jurisdiction` and `rights_note` out into separate objects feels like a bit of an anti-pattern to me. I get that you're trying to avoid repetition, but I feel like the additional complexity added by creating another object definition (which has only one field associated with it) outweighs anything that might be gained by not just defining that field.
- Similarly, I would suggest a slightly different pattern for the rights basis handling, which follows this pattern: https://json-schema.org/understanding-json-schema/structuring.html#extending, which I think might help simplify things and get rid of some of those conditionals.
  - Create a basic rights_basis object and then extend that with each of the four specific bases fields.
  - In the rights_statement, validated against a `oneOf` array of the four specific bases. 
- We're missing a fixture for an invalid `basis_note` on a rights statement
- This is a bit of a broader question, but given our approach with rights statements in Aquila, I'm wondering if a) `rights_identifier` should just be `id` and b) if we need a `uri` field that has a ref to the rights statement. Thoughts, @bonniegee?

**Comment by bonniegee** [comment]
>     * This is a bit of a broader question, but given our approach with rights statements in Aquila, I'm wondering if a) `rights_identifier` should just be `id` and b) if we need a `uri` field that has a ref to the rights statement. Thoughts, @bonniegee?

I'm a little unclear at what the `rights_identifier` is (i.e., what system would be generating it), and what its purpose is. Aquila has IDs for rights shells for the purpose of being able to minimize the amount of text/policy that needs to be approved, but those IDs are not meant to persist with the PREMIS statements that have been generated. Right now, this PREMIS information will be used by two systems:
1. Fornax (which transforms this information for Archivematica)
2. The IIIIF conversion pipeline
I don't think either of these would need identifiers associated with rights statements?

**Review by bonniegee — CHANGES_REQUESTED** [review]
Looks good! A couple of things:

- One of the valid rights json fixtures should have a blank rights granted array.
- Why is the `note` an array instead of a string?

**Comment by p-galligan** [comment]
> Looks good! A couple of things:
> 
> * One of the valid rights json fixtures should have a blank rights granted array.
> * Why is the `note` an array instead of a string?

@bonniegee I'm still making changes, but are rights statements allowed to have empty rights granted information? My understanding was that rights granted was mandatory. If it's not, I'll change around some required fields.

In the work I'm doing, there will be a basis_note, and granted_note. According to PREMIS, the basis and granted notes are marked as repeatable in PREMIS, which suggests that an array would be appropriate here. However, if we don't think we'll ever be able to have an array, I can change those to strings.

**Comment by p-galligan** [comment]
Ok @helrond I haven't touched the rights identifier or notes yet because I'm not sure what direction we want to go for those. Either way would be pretty simple IMO.

I think the usage of the rights identifier would be for the rightsStatementIdentifier field as explained in the PREMIS specification. However, if we're sure we won't ever need an identifier to persist across systems for individual rights statements, I can get rid of it.

**Comment by helrond** [comment]
This is looking really nice - I think those last changes really improved the readability and maintainability of the schema.

Regarding notes, I'd suggest just going with a string rather than an array. In general we try and steer folks away from lengthy notes because these rights statements are supposed to be machine-readable, not human readable.

I'd also suggest yanking out the `rights_identifier` for now. No sense in having a field in the schema that we're not using.

**Review by helrond — APPROVED** [review]

**Review by bonniegee — APPROVED** [review]
✨
